src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,769 @@
+  bool init_timers();
+  void stop_timer();
+
+  void trigger_async_processing_of_cpu_time_jfr_requests();
+};
+
+JfrCPUSamplerThread::JfrCPUSamplerThread(double rate, bool auto_adapt) :
+  _sample(),
+  _sampler_thread(nullptr),
+  _rate(rate),
+  _auto_adapt(auto_adapt),
+  _current_sampling_period_ns(compute_sampling_period(rate)),
+  _disenrolled(true),
+  _active_signal_handlers(STOP_SIGNAL_BIT),
+  _is_async_processing_of_cpu_time_jfr_requests_triggered(false),
+  _warned_about_timer_creation_failure(false),
+  _signal_handler_installed(false) {
+  assert(rate >= 0, "invariant");
+}
+
+void JfrCPUSamplerThread::trigger_async_processing_of_cpu_time_jfr_requests() {
+  Atomic::release_store(&_is_async_processing_of_cpu_time_jfr_requests_triggered, true);
+}
+
+void JfrCPUSamplerThread::on_javathread_create(JavaThread* thread) {
+  if (thread->is_hidden_from_external_view() || thread->is_JfrRecorder_thread() ||
+      !Atomic::load_acquire(&_signal_handler_installed)) {
+    return;
+  }
+  JfrThreadLocal* tl = thread->jfr_thread_local();
+  assert(tl != nullptr, "invariant");
+  tl->cpu_time_jfr_queue().resize_for_period(_current_sampling_period_ns / 1000000);
+  timer_t timerid;
+  if (create_timer_for_thread(thread, timerid)) {
+    tl->set_cpu_timer(&timerid);
+  } else {
+    if (!Atomic::or_then_fetch(&_warned_about_timer_creation_failure, true)) {
+      log_warning(jfr)("Failed to create timer for a thread");
+    }
+    tl->deallocate_cpu_time_jfr_queue();
+  }
+}
+
+void JfrCPUSamplerThread::on_javathread_terminate(JavaThread* thread) {
+  JfrThreadLocal* tl = thread->jfr_thread_local();
+  assert(tl != nullptr, "invariant");
+  timer_t* timer = tl->cpu_timer();
+  if (timer == nullptr) {
+    return; // no timer was created for this thread
+  }
+  tl->unset_cpu_timer();
+  tl->deallocate_cpu_time_jfr_queue();
+  s4 lost_samples = tl->cpu_time_jfr_queue().lost_samples();
+  if (lost_samples > 0) {
+    JfrCPUTimeThreadSampling::send_lost_event(JfrTicks::now(), JfrThreadLocal::thread_id(thread), lost_samples);
+  }
+}
+
+void JfrCPUSamplerThread::start_thread() {
+  if (os::create_thread(this, os::os_thread)) {
+    os::start_thread(this);
+  } else {
+    log_error(jfr)("Failed to create thread for thread sampling");
+  }
+}
+
+void JfrCPUSamplerThread::enroll() {
+  if (Atomic::cmpxchg(&_disenrolled, true, false)) {
+    Atomic::store(&_warned_about_timer_creation_failure, false);
+    allow_signal_handlers();
+    log_trace(jfr)("Enrolling CPU thread sampler");
+    _sample.signal();
+    if (!init_timers()) {
+      log_error(jfr)("Failed to initialize timers for CPU thread sampler");
+      disenroll();
+      return;
+    }
+    log_trace(jfr)("Enrolled CPU thread sampler");
+  }
+}
+
+void JfrCPUSamplerThread::disenroll() {
+  if (!Atomic::cmpxchg(&_disenrolled, false, true)) {
+    log_trace(jfr)("Disenrolling CPU thread sampler");
+    if (Atomic::fetch_then_and(&_signal_handler_installed, false)) {
```

Comment:
fetch_then_and with false? Must be simpler way to express this? Like Atomic::load(&_signal_handler_installed)?

Reply:
I also want to set it. Wanted to do a simple ::xchg, but this doesn't exist for booleans.

Reply:
You're right, it's not required. I removed it.